File: app/src/main/java/org/kindone/willingtodo/pomodorotimer/StopWatch.java
package org.kindone.willingtodo.pomodorotimer;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * Created by kindone on 2017. 1. 11..
 */

public class StopWatch implements TimerContext{

    private TimerState mTimerState;

    private long mDurationMs;

    private ScheduledFuture<?> mTickerHandle;

    private TimerEventListener mTimerEventListener = null;

    private final long BUFFER_TIME_MS = 1000;

    private final ScheduledExecutorService scheduler =
            Executors.newScheduledThreadPool(1);


    // advanced timer events
    public StopWatch(TimerEventListener timerEventListener) {
        setInitialTimerState();
        mTimerEventListener = timerEventListener;
    }

    // simple tick event only
    public StopWatch(Runnable onTick) {
        setInitialTimerState();
        mTimerEventListener = createTimerEventListener(onTick);
    }

    public void startTimer(long newDurationMs) {
        mDurationMs = newDurationMs;
        boolean changedState = setRunningTimerState();
        startBackgroundTicking(mDurationMs + BUFFER_TIME_MS);

        if(changedState)
            dispatchTimerStartedEvent();
    }

    public void restartTimer(long newDurationMs, long newStartedTimeMs) {
        setRunningTimerState(newDurationMs, newStartedTimeMs);
        mDurationMs = getRemainingTimeMs();
        startBackgroundTicking(mDurationMs + BUFFER_TIME_MS);
        dispatchTimerStartedEvent();
    }

    public void pauseTimer() {
        boolean changedState = setPausedTimerState();
        if(changedState)
            dispatchTimerStoppedEvent();
    }

    public void pauseTimer(long newRemainingTimeMs) {
        boolean changedState = setPausedTimerState(newRemainingTimeMs);
        if(changedState)
            dispatchTimerStoppedEvent();
    }

    public void resumeTimer() {
        boolean changedState = setResumedTimerState();
        startBackgroundTicking(getRemainingTimeMs() + BUFFER_TIME_MS);

        if(changedState)
            dispatchTimerStartedEvent();
    }

    public void stopTimer() {
        boolean changedState = setStoppedTimerState();
        if(changedState)
            dispatchTimerStoppedEvent();
    }

    public void cleanUp() {
        stopTimer();
        stopBackgroundTicking();
    }



    public boolean isRunning() { return mTimerState.isRunning();}

    public boolean isStopped() { return mTimerState.isStopped();}



    public long getDurationMs() {
        return mDurationMs;
    }

    public long getStartedTimeMs() {
        return mTimerState.getStartedTimeMs();
    }

    public long getRemainingTimeMs() {
        return mTimerState.getRemainingTimeMs();
    }

    public String getRemainingTimeStr() {
        String str;
        synchronized (mTimerState) {
            str = mTimerState.getRemainingTimeStr();
        }
        return str;
    }



    @Override
    public void changeState(TimerState state) {
        synchronized (mTimerState) {
            mTimerState = state;
        }
    }



    private void setInitialTimerState()
    {
        mTimerState = new TimerInitial(this);
    }

    private void setRunningTimerState(long durationMs, long startedTimeMs)
    {
        mTimerState = new TimerRunning(this, durationMs, startedTimeMs);
    }

    private boolean setRunningTimerState()
    {
        return mTimerState.start();
    }

    private boolean setPausedTimerState(long newRemainingTimeMs)
    {
        boolean wasRunning = mTimerState.isRunning();
        mTimerState = new TimerPaused(this, newRemainingTimeMs);
        return wasRunning;
    }

    private boolean setPausedTimerState()
    {
        return mTimerState.pause();
    }

    private boolean setResumedTimerState()
    {
        return mTimerState.resume();
    }

    private boolean setStoppedTimerState()
    {
        return mTimerState.stop();
    }


    private void startBackgroundTicking(long maxDurationMs) {
        stopBackgroundTicking();
        mTickerHandle = scheduleTimerTick();
        scheduleTickerExpiry(maxDurationMs);
    }

    private void stopBackgroundTicking() {
        if(isTickerActive()) {
            cancelTicker();
        }
    }



    private ScheduledFuture<?> scheduleTimerTick()
    {
        Runnable scheduledJob = new Ticker();
        long initialDelay = 500;
        long period = 500;
        return scheduler.scheduleAtFixedRate(scheduledJob, initialDelay, period, TimeUnit.MILLISECONDS);
    }

    private void scheduleTickerExpiry(long maxDurationMs)
    {
        Runnable scheduledJob = new ExpireTick();
        scheduler.schedule(scheduledJob, maxDurationMs, TimeUnit.MILLISECONDS);
    }



    private boolean isTickerActive()
    {
        return mTickerHandle != null && !mTickerHandle.isCancelled();
    }

    private void cancelTicker()
    {
        mTickerHandle.cancel(true);
    }



    private void dispatchTimerStartedEvent()
    {
        mTimerEventListener.onTimerStarted();
    }

    private void dispatchTimerStoppedEvent()
    {
        mTimerEventListener.onTimerStopped();
    }

    private void dispatchTimerTickEvent()
    {
        mTimerEventListener.onTick();
    }

    private void dispatchTimerExpiredEvent()
    {
        mTimerEventListener.onTimerExpired();
    }


    private TimerEventListener createTimerEventListener(final Runnable runOnTick)
    {
        return new TimerEventListener() {

            @Override
            public void onTick() {
                runOnTick.run();
            }

            @Override
            public void onTimerStarted() { }

            @Override
            public void onTimerStopped() { }

            @Override
            public void onTimerExpired() { }
        };
    }


    public interface TimerEventListener {
        void onTick();

        void onTimerStarted();

        void onTimerStopped();

        void onTimerExpired();
    }


    class Ticker implements Runnable {
        public void run() {
            if(!isRunning()) {
                stopBackgroundTicking();
            }
            if(getRemainingTimeMs() <= 0)
            {
                stopTimer();
            }

            dispatchTimerTickEvent();
        }
    }


    class ExpireTick implements Runnable {
        public void run() {
            stopBackgroundTicking();
            dispatchTimerExpiredEvent();
        }
    }
}

File: app/src/main/java/org/kindone/willingtodo/pomodorotimer/StopWatch.kt
package org.kindone.willingtodo.pomodorotimer

import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * Created by kindone on 2017. 1. 11..
 */

class StopWatch : TimerContext {

    private var mTimerState: TimerState? = null

    private var mDurationMs: Long = 0

    private var mTickerHandle: ScheduledFuture<*>? = null

    private var mTimerEventListener: TimerEventListener? = null

    private val BUFFER_TIME_MS: Long = 1000

    private val scheduler = Executors.newScheduledThreadPool(1)


    // advanced timer events
    constructor(timerEventListener: TimerEventListener) {
        setInitialTimerState()
        mTimerEventListener = timerEventListener
    }

    // simple tick event only
    constructor(onTick: Runnable) {
        setInitialTimerState()
        mTimerEventListener = createTimerEventListener(onTick)
    }

    fun startTimer(newDurationMs: Long) {
        mDurationMs = newDurationMs
        val changedState = setRunningTimerState()
        startBackgroundTicking(mDurationMs + BUFFER_TIME_MS)

        if (changedState)
            dispatchTimerStartedEvent()
    }

    fun restartTimer(newDurationMs: Long, newStartedTimeMs: Long) {
        setRunningTimerState(newDurationMs, newStartedTimeMs)
        mDurationMs = remainingTimeMs
        startBackgroundTicking(mDurationMs + BUFFER_TIME_MS)
        dispatchTimerStartedEvent()
    }

    fun pauseTimer() {
        val changedState = setPausedTimerState()
        if (changedState)
            dispatchTimerStoppedEvent()
    }

    fun pauseTimer(newRemainingTimeMs: Long) {
        val changedState = setPausedTimerState(newRemainingTimeMs)
        if (changedState)
            dispatchTimerStoppedEvent()
    }

    fun resumeTimer() {
        val changedState = setResumedTimerState()
        startBackgroundTicking(remainingTimeMs + BUFFER_TIME_MS)

        if (changedState)
            dispatchTimerStartedEvent()
    }

    fun stopTimer() {
        val changedState = setStoppedTimerState()
        if (changedState)
            dispatchTimerStoppedEvent()
    }

    fun cleanUp() {
        stopTimer()
        stopBackgroundTicking()
    }


    val isRunning: Boolean
        get() = mTimerState!!.isRunning()

    val isStopped: Boolean
        get() = mTimerState!!.isStopped()


    override fun getDurationMs(): Long {
        return mDurationMs
    }

    val startedTimeMs: Long
        get() = mTimerState!!.getStartedTimeMs()

    val remainingTimeMs: Long
        get() = mTimerState!!.getRemainingTimeMs()

    val remainingTimeStr: String
        get() {
            var str: String =
            synchronized(mTimerState!!) {
                mTimerState!!.getRemainingTimeStr()
            }
            return str
        }


    override fun changeState(state: TimerState) {
        synchronized(mTimerState!!) {
            mTimerState = state
        }
    }


    private fun setInitialTimerState() {
        mTimerState = TimerInitial(this)
    }

    private fun setRunningTimerState(durationMs: Long, startedTimeMs: Long) {
        mTimerState = TimerRunning(this, durationMs, startedTimeMs)
    }

    private fun setRunningTimerState(): Boolean {
        return mTimerState!!.start()
    }

    private fun setPausedTimerState(newRemainingTimeMs: Long): Boolean {
        val wasRunning = mTimerState!!.isRunning()
        mTimerState = TimerPaused(this, newRemainingTimeMs)
        return wasRunning
    }

    private fun setPausedTimerState(): Boolean {
        return mTimerState!!.pause()
    }

    private fun setResumedTimerState(): Boolean {
        return mTimerState!!.resume()
    }

    private fun setStoppedTimerState(): Boolean {
        return mTimerState!!.stop()
    }


    private fun startBackgroundTicking(maxDurationMs: Long) {
        stopBackgroundTicking()
        mTickerHandle = scheduleTimerTick()
        scheduleTickerExpiry(maxDurationMs)
    }

    private fun stopBackgroundTicking() {
        if (isTickerActive) {
            cancelTicker()
        }
    }


    private fun scheduleTimerTick(): ScheduledFuture<*> {
        val scheduledJob = Ticker()
        val initialDelay: Long = 500
        val period: Long = 500
        return scheduler.scheduleAtFixedRate(scheduledJob, initialDelay, period, TimeUnit.MILLISECONDS)
    }

    private fun scheduleTickerExpiry(maxDurationMs: Long) {
        val scheduledJob = ExpireTick()
        scheduler.schedule(scheduledJob, maxDurationMs, TimeUnit.MILLISECONDS)
    }


    private val isTickerActive: Boolean
        get() = mTickerHandle != null && !mTickerHandle!!.isCancelled

    private fun cancelTicker() {
        mTickerHandle!!.cancel(true)
    }


    private fun dispatchTimerStartedEvent() {
        mTimerEventListener!!.onTimerStarted()
    }

    private fun dispatchTimerStoppedEvent() {
        mTimerEventListener!!.onTimerStopped()
    }

    private fun dispatchTimerTickEvent() {
        mTimerEventListener!!.onTick()
    }

    private fun dispatchTimerExpiredEvent() {
        mTimerEventListener!!.onTimerExpired()
    }


    private fun createTimerEventListener(runOnTick: Runnable): TimerEventListener {
        return object : TimerEventListener {

            override fun onTick() {
                runOnTick.run()
            }

            override fun onTimerStarted() {}

            override fun onTimerStopped() {}

            override fun onTimerExpired() {}
        }
    }


    interface TimerEventListener {
        fun onTick()

        fun onTimerStarted()

        fun onTimerStopped()

        fun onTimerExpired()
    }


    internal inner class Ticker : Runnable {
        override fun run() {
            if (!isRunning) {
                stopBackgroundTicking()
            }
            if (remainingTimeMs <= 0) {
                stopTimer()
            }

            dispatchTimerTickEvent()
        }
    }


    internal inner class ExpireTick : Runnable {
        override fun run() {
            stopBackgroundTicking()
            dispatchTimerExpiredEvent()
        }
    }
}
